File: packages/arcantry/src/managedContent.ts
export const managedSectionStart = '<!-- arcantry:start -->';
export const managedSectionEnd = '<!-- arcantry:end -->';

export type ManagedSectionResult =
  | { status: 'unchanged'; content: string }
  | { status: 'changed'; content: string }
  | { status: 'conflict'; content: string; reason: string };

export const renderManagedSection = (body: string): string => `${managedSectionStart}\n${body.trim()}\n${managedSectionEnd}`;

export const upsertManagedSection = (existing: string, body: string): ManagedSectionResult => {
  const desired = renderManagedSection(body);
  const startIndexes = findAll(existing, managedSectionStart);
  const endIndexes = findAll(existing, managedSectionEnd);

  if (startIndexes.length === 0 && endIndexes.length === 0) {
    const prefix = existing.length === 0 ? '' : existing.endsWith('\n') ? '\n' : '\n\n';
    return { status: 'changed', content: `${existing}${prefix}${desired}\n` };
  }

  if (startIndexes.length !== 1 || endIndexes.length !== 1 || startIndexes[0] > endIndexes[0]) {
    return { status: 'conflict', content: existing, reason: 'Arcantry section markers are incomplete or duplicated.' };
  }

  const sectionEnd = endIndexes[0] + managedSectionEnd.length;
  const current = existing.slice(startIndexes[0], sectionEnd);
  if (current === desired) {
    return { status: 'unchanged', content: existing };
  }

  return {
    status: 'changed',
    content: `${existing.slice(0, startIndexes[0])}${desired}${existing.slice(sectionEnd)}`,
  };
};

export const removeManagedSection = (existing: string): ManagedSectionResult => {
  const startIndexes = findAll(existing, managedSectionStart);
  const endIndexes = findAll(existing, managedSectionEnd);

  if (startIndexes.length === 0 && endIndexes.length === 0) {
    return { status: 'unchanged', content: existing };
  }

  if (startIndexes.length !== 1 || endIndexes.length !== 1 || startIndexes[0] > endIndexes[0]) {
    return { status: 'conflict', content: existing, reason: 'Arcantry section markers are incomplete or duplicated.' };
  }

  const sectionEnd = endIndexes[0] + managedSectionEnd.length;
  const before = existing.slice(0, startIndexes[0]).replace(/\n+$/, '');
  const after = existing.slice(sectionEnd).replace(/^\n+/, '');
  const content = [before, after].filter((part) => part.length > 0).join('\n\n');

  return { status: 'changed', content: content.length > 0 ? `${content}\n` : '' };
};

export const containsManagedSection = (content: string): boolean =>
  content.includes(managedSectionStart) && content.includes(managedSectionEnd);

const findAll = (content: string, marker: string): number[] => {
  const indexes: number[] = [];
  let fromIndex = 0;
  while (fromIndex < content.length) {
    const index = content.indexOf(marker, fromIndex);
    if (index === -1) {
      break;
    }
    indexes.push(index);
    fromIndex = index + marker.length;
  }
  return indexes;
};

File: crates/arcantry-core/src/managed_content.rs
pub const MANAGED_SECTION_START: &str = "<!-- arcantry:start -->";
pub const MANAGED_SECTION_END: &str = "<!-- arcantry:end -->";

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ManagedSectionResult {
  Unchanged(String),
  Changed(String),
  Conflict { content: String, reason: String },
}

pub fn render_managed_section(body: &str) -> String {
  format!(
    "{MANAGED_SECTION_START}\n{}\n{MANAGED_SECTION_END}",
    body.trim()
  )
}

pub fn upsert_managed_section(existing: &str, body: &str) -> ManagedSectionResult {
  let desired = render_managed_section(body);
  let starts: Vec<_> = existing
    .match_indices(MANAGED_SECTION_START)
    .map(|(index, _)| index)
    .collect();
  let ends: Vec<_> = existing
    .match_indices(MANAGED_SECTION_END)
    .map(|(index, _)| index)
    .collect();
  if starts.is_empty() && ends.is_empty() {
    let prefix = if existing.is_empty() {
      ""
    } else if existing.ends_with('\n') {
      "\n"
    } else {
      "\n\n"
    };
    return ManagedSectionResult::Changed(format!("{existing}{prefix}{desired}\n"));
  }
  if starts.len() != 1 || ends.len() != 1 || starts[0] > ends[0] {
    return ManagedSectionResult::Conflict {
      content: existing.to_owned(),
      reason: "Arcantry section markers are incomplete or duplicated.".to_owned(),
    };
  }
  let section_end = ends[0] + MANAGED_SECTION_END.len();
  if existing[starts[0]..section_end] == *desired {
    return ManagedSectionResult::Unchanged(existing.to_owned());
  }
  ManagedSectionResult::Changed(format!(
    "{}{}{}",
    &existing[..starts[0]],
    desired,
    &existing[section_end..]
  ))
}

pub fn remove_managed_section(existing: &str) -> ManagedSectionResult {
  let starts: Vec<_> = existing
    .match_indices(MANAGED_SECTION_START)
    .map(|(index, _)| index)
    .collect();
  let ends: Vec<_> = existing
    .match_indices(MANAGED_SECTION_END)
    .map(|(index, _)| index)
    .collect();
  if starts.is_empty() && ends.is_empty() {
    return ManagedSectionResult::Unchanged(existing.to_owned());
  }
  if starts.len() != 1 || ends.len() != 1 || starts[0] > ends[0] {
    return ManagedSectionResult::Conflict {
      content: existing.to_owned(),
      reason: "Arcantry section markers are incomplete or duplicated.".to_owned(),
    };
  }
  let section_end = ends[0] + MANAGED_SECTION_END.len();
  let before = existing[..starts[0]].trim_end_matches('\n');
  let after = existing[section_end..].trim_start_matches('\n');
  let body = [before, after]
    .into_iter()
    .filter(|part| !part.is_empty())
    .collect::<Vec<_>>()
    .join("\n\n");
  ManagedSectionResult::Changed(if body.is_empty() {
    String::new()
  } else {
    format!("{body}\n")
  })
}

pub fn contains_managed_section(content: &str) -> bool {
  content.contains(MANAGED_SECTION_START) && content.contains(MANAGED_SECTION_END)
}
